### ISSUE TITLE
Updating model weights using custom formula

### ISSUE THREAD

**Issue by Mrugankakarte** [rstudio/keras#618]
Hi,

I am trying to implement the Online Deep learning model from [this paper](https://arxiv.org/pdf/1711.03705.pdf). 

As mentioned in the paper I would like to update the model weights as mentioned on page 4 equation 3. Does keras updates the weights in similar fashion or I need to do some changes?

For example,
Suppose I have a 3 layer NNmodel. According to paper, weights for layer 2 are to be updated only using back-propagation from layer 2&3.

PS: I asked the question to author, their solution only supports keras 1.x in which they modified the training.py

**Comment by skeydan** [comment]
Well, looking at the paper it _does_ seem they do some custom stuff to backprop here (although if they did it in `training.py`, it must be less "deep down" than it sounds (as it was possible for them to do it in keras instead of the backend)).

You could try asking them if they're interested in updating their `training.py` to current keras...

**Comment by Mrugankakarte** [comment]
Hi @skeydan ,

Just wanted to confirm a thing. 

In multi-output model, for example...
```
library(keras)

train <- matrix(data = rnorm(20), nrow = 1000, ncol = 28)
y_train <- to_categorical(sample(c(0,1), size = 1000, replace = T), num_classes = 2)

inp <- layer_input(shape = c(28), name = 'input')
layer0_out <- inp %>% layer_dense(units = 2, activation = 'softmax', name = 'layer0_out')

layer1 <- inp %>% layer_dense(units = 15, activation = 'relu')
layer1_out <- layer1 %>% layer_dense(units = 2, activation = 'softmax', name = 'layer1_out')

layer2 <- layer1 %>% layer_dense(units = 15, activation = 'relu')
layer2_out <- layer2 %>% layer_dense(units = 2, activation = 'softmax', name = 'layer2_out')

layer3 <- layer2 %>% layer_dense(units = 15, activation = 'relu')
layer3_out <- layer3 %>% layer_dense(units = 2, activation = 'softmax', name = 'layer3_out')


model  <- keras_model(inputs = inp, outputs = list(layer0_out, layer1_out, layer2_out, layer3_out))
model %>% summary()
model %>% compile(optimizer = optimizer_adam(), 
                  loss = 'categorical_crossentropy',
                  metrics = 'accuracy')
model %>% fit(x = train, y = list(y_train, y_train, y_train, y_train),
              epochs = 1, batch_size = 1, shuffle = F) 
```
during back-propagation in keras....the weights for layer1 (assume w1) are update as 
```
loss1 = categorical_crossentropy(layer1_out, y_train)
loss2 = categorical_crossentropy(layer2_out, y_train) 
loss3 = categorical_crossentropy(layer3_out, y_train)

dw1 = gradient(w1 with respect to loss1) + gradient(w1 with respect to loss2) +gradient(w1 with respect to loss3)
dw2 = gradient(w2 with respect to loss2) +gradient(w2 with respect to loss3)
```
is my interpretation correct?

**Comment by skeydan** [comment]
Let's say I would share your understanding it should work like that :-)

I didn't have a chance yet to look at the above code, - in any case a question: Is that your port of what you saw them do in Python (meaning: did they implement their logic as a callback in Python, too)?
Or did they do it otherwise, and trying a callback is your workaround?

**Comment by Mrugankakarte** [comment]
Thanks. In that case the only problem is how to multiply the gradients with alpha before updating the weights. Is it possible to do so using callback, maybe using `model$get_weights()` and `model$set_weights()`? Or update the `loss_weights` parameter in model from callback so that it will carried back while back-propagating?

Callback is being used to update the values of `alpha` and the authors have done the same using callback [link to their code](https://github.com/LIBOL/ODL/blob/master/src/train.py). As of now I haven't done any workaround.

I tried comparing their [training.py](https://github.com/LIBOL/ODL/blob/master/training.py) with the original [training.py](https://github.com/keras-team/keras/blob/master/keras/engine/training.py), but couldn't understand the exact difference. My guess is that the authors of paper have modified the original file to add a parameter `hedge` ([this can be seen in `model.compile()`](https://github.com/LIBOL/ODL/blob/master/src/train.py))which acts a switch between original and their implementation.

**Comment by skeydan** [comment]
I don't have the time to analyse the details, but from the files you're linking 

- their test code, plus callback (https://github.com/LIBOL/ODL/blob/master/src/train.py)
- their modified `training.py` (https://github.com/LIBOL/ODL/blob/master/training.py)

it seems to me that they introduce a new model variable called `holder`, which has to be updated from the callback. Thus probably, just a callback alone won't work. (You can't just update `weights` instead.)

However, I don't know if one could achieve the same goals without that `holder` - at least the authors must have deemed its introduction necessary. So my suggestions would be, either you have an idea to do the same thing differently from them, which does not require modifying the Python code, or you might try porting their changes to current Keras.

In the latter case, what I'd do would be starting with the Python only, that is, trying if their test code runs with your new `training.py`. And only if that works, add in the R callback.

**Comment by Mrugankakarte** [comment]
So the idea that they implemented is `total_loss = alpha(hedge weights) * losses` and while back-propagating these `hedge weights` would be automatically carried back while updating the model weights. Just got it confirmed from the authors.

Isn't the `loss_weights` parameter doing the same thing?

**Comment by skeydan** [comment]
what do you mean by 

> while back-propagating these hedge weights would be automatically carried back while updating the model weights

?
Are they being updated / modified somehow?

**Comment by Mrugankakarte** [comment]
> Edit: @skeydan You are right about the `holder` part. I got this reply from authors when I asked them about the changes in `training.py` file
> 
> 
> line 678-684 
>```
>self.holder = []
>
>       if hedge == True:
>            for i in range(len(loss_weights)):
>                self.holder.append(K.placeholder(ndim = 0)) 
>
>            total_loss = K.sum(K.dot(self.holder, self.metrics_tensors))
>```
>where the total loss is computed as a dot product between the hedge weight vector and the loss vector (self.holder is the hedge weights , self.metrics_tensors is the losses).
> So the remaining step is to set the hedge weights before each computation in training and testing, which you can look at line 922 (training), line 1021 (testing).
> 

The values are updated using callback then multiplied with losses to get `total_loss`. And when we take the derivatives for back-propagation; `alpha` is already present, so no need to multiply the model weights with `alpha` again.

**Comment by skeydan** [comment]
that sounds good!

**Comment by Mrugankakarte** [comment]
So isn't the `loss_weights` parameter doing the same thing? I was thinking of updating the `loss_weights` from callback as 
```
on_batch_end = function(batch, logs = list()) {
                                    [update the values of alpha as alpha = beta^losses](as mentioned in paper)
                                    self$model$loss_weights = alpha     
                                 }
```
instead of creating a new variable `holder` and modifying `training.py`. This can solve the problem right?

So far it working as expected on test data, but still have to test on some benchmark dataset.

**Comment by skeydan** [comment]
Hi, did you get this to work so I could close the issue?

**Comment by skeydan** [comment]
cool, happy to hear that!
